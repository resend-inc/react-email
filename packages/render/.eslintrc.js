module.exports = {
  extends: ["custom/react-internal"],
  rules: {
    "import/no-default-export": "off",
    "@typescript-eslint/consistent-type-imports": "off",
    "tsdoc/syntax": "off"
  },
};
