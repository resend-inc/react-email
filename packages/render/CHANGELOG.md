# @react-email/render

## 0.0.17
