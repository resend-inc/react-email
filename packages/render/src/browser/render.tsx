import { convert } from "html-to-text";
import { Suspense } from "react";
import { renderToReadableStream } from "react-dom/server";
import { Options } from "../shared/options";
import { plainTextSelectors } from "../shared/plain-text-selectors";
import { pretty } from "../shared/pretty";

async function streamToString(
  stream: ReadableStream<Uint8Array>,
): Promise<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder("utf-8");
  let result = "";
  let done: boolean | undefined;

  while (!done) {
    // eslint-disable-next-line no-await-in-loop
    const { value, done: doneReading } = await reader.read();
    done = doneReading;
    if (value) {
      result += decoder.decode(value, { stream: !done });
    }
  }

  return result;
}

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
<<<<<<< HEAD
  const reactDOMServer = await import("react-dom/server");

  let html!: string;
  if (Object.hasOwn(reactDOMServer, "renderToReadableStream")) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }
=======
  const readableStream = await renderToReadableStream(suspendedElement);
  const html = await streamToString(readableStream);
>>>>>>> 447c7780 (refactor: `render` package)

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
