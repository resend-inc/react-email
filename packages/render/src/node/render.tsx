import { Writable } from "node:stream";
import { renderToPipeableStream } from "react-dom/server";
import { Suspense } from "react";
import { convert } from "html-to-text";
import { Options } from "../shared/options";
import { pretty } from "../shared/pretty";
import { plainTextSelectors } from "../shared/plain-text-selectors";

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const html = await new Promise<string>((resolve, reject) => {
    let htmlChunksString = "";

    const writable = new Writable({
      write(chunk: Buffer, _encoding, callback) {
        htmlChunksString += chunk.toString();
        callback();
      },
    });

    writable.on("finish", () => {
      resolve(htmlChunksString);
    });

    writable.on("error", (err) => {
      reject(err);
    });

    const stream = renderToPipeableStream(suspendedElement, {
      onAllReady() {
        stream.pipe(writable);
      },
      onError(error) {
        reject(error as Error);
      },
    });
  });

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
