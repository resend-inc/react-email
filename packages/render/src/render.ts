import * as ReactDomServer from "react-dom/server";
import { convert } from "html-to-text";
import pretty from "pretty";

export interface Options {
  pretty?: boolean;
  plainText?: boolean;
}

export const render = (component: React.ReactElement, options?: Options) => {
  const markup = ReactDomServer.renderToStaticMarkup(component);
  if (options?.plainText) {
    return convertToPlainText(markup);
  }
  
  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const document = `${doctype}${markup}`;

  if (options && options.pretty) {
    return pretty(document);
  }

  return document;
};

export const convertToPlainText = (staticMarkup: string) => {
  return convert(staticMarkup, {
    selectors: [
      { selector: "img", format: "skip" },
      { selector: "#__react-email-preview", format: "skip" },
    ],
  });
};