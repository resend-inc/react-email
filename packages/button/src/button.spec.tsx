import { render } from "@react-email/render";
import { Button } from "./index";

describe("<Button> component", () => {
  it("renders children correctly", () => {
    const testMessage = "Test message";
    const html = render(<Button>{testMessage}</Button>);
    expect(html).toContain(testMessage);
  });

  it("passes style and other props correctly", () => {
    const style = { backgroundColor: "red" };
    const html = render(
      <Button data-testid="button-test" style={style}>
        Test
      </Button>,
    );
    expect(html).toContain("background-color:red");
    expect(html).toContain('data-testid="button-test"');
  });

  it("renders correctly  with padding values from style prop", () => {
    const actualOutput = render(
      <Button href="https://example.com" style={{ padding: "12px 20px" }} />,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it("renders the <Button> component with no padding value", () => {
    const actualOutput = render(<Button href="https://example.com" />);
    expect(actualOutput).toMatchSnapshot();
  });

  it("should allow users to overwrite style props", () => {
    const actualOutput = render(
      <Button
        style={{
          lineHeight: "150%",
          display: "block",
          textDecoration: "underline red",
          maxWidth: "50%",
        }}
      />,
    );
    expect(actualOutput).toMatchSnapshot();
  });
});
