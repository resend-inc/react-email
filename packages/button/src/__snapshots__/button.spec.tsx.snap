// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders the <Button> component 1`] = `"<!DOCTYPE htmlPUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><a href=\\"https://example.com\\" target=\\"_blank\\" style=\\"line-height:100%;text-decoration:none;display:inline-block;max-width:100%;padding:0px 0px\\"><span><!--[if mso]><i style=\\"letter-spacing: undefinedpx;mso-font-width:-100%;mso-text-raise:0\\" hidden>&nbsp;</i><![endif]--></span><span style=\\"max-width:100%;display:inline-block;line-height:120%;text-decoration:none;text-transform:none;mso-padding-alt:0px;mso-text-raise:0\\"></span><span><!--[if mso]><i style=\\"letter-spacing: undefinedpx;mso-font-width:-100%\\" hidden>&nbsp;</i><![endif]--></span></a>"`;
